ArgsParsing.Tests/TypeParsersTest.cs

```diff
@@ -0,0 +1,165 @@
+﻿using System;
+using System.Collections.Immutable;
+using System.Threading.Tasks;
+using ArgsParsing.TypeParsers;
+using ArgsParsing.Types;
+using Moq;
+using NUnit.Framework;
+using Persistence.Models;
+using Persistence.Repos;
+
+namespace ArgsParsing.Tests
+{
+    public class TypeParsersTest
+    {
+        [Test]
+        public async Task TestAnyOrderParser()
+        {
+            var argsParser = new ArgsParser();
+            argsParser.AddArgumentParser(new AnyOrderParser(argsParser));
+            argsParser.AddArgumentParser(new StringParser());
+            argsParser.AddArgumentParser(new IntParser());
+
+            var args1 = ImmutableList.Create("123", "foo");
+            var args2 = ImmutableList.Create("foo", "123");
+            (int int1, string string1) = (await argsParser.Parse<AnyOrder<int, string>>(args1)).AsTuple();
+            (int int2, string string2) = (await argsParser.Parse<AnyOrder<int, string>>(args2)).AsTuple();
+            Assert.AreEqual(123, int1);
+            Assert.AreEqual(123, int2);
+            Assert.AreEqual("foo", string1);
+            Assert.AreEqual("foo", string2);
+
+            var ex = Assert.ThrowsAsync<ArgsParseFailure>(() => argsParser
+                .Parse<AnyOrder<int, string>>(ImmutableList.Create("foo", "bar")));
+            Assert.AreEqual("did not recognize 'foo' as a number", ex.Message);
+        }
+
+        [Test]
+        public async Task TestDateTimeParser()
+        {
+            var argsParser = new ArgsParser();
+            argsParser.AddArgumentParser(new DateTimeUtcParser());
+
+            var result1 =
+                await argsParser.Parse<DateTime>(args: ImmutableList.Create("2020-03-22", "01:59:20Z"));
+            var result2 = await argsParser.Parse<DateTime>(args: ImmutableList.Create("2020-03-22T01:59:20Z"));
+
+            DateTime refDateTime = DateTime.SpecifyKind(DateTime.Parse("2020-03-22 01:59:20+00"), DateTimeKind.Utc);
+            Assert.AreEqual(refDateTime, result1);
+            Assert.AreEqual(refDateTime, result2);
+            Assert.AreEqual(result1, result2);
+
+            var ex1 = Assert.ThrowsAsync<ArgsParseFailure>(() => argsParser
+                .Parse<DateTime>(ImmutableList.Create("2020-03-22T01:59:20+02")));
+            var ex2 = Assert.ThrowsAsync<ArgsParseFailure>(() => argsParser
+                .Parse<DateTime>(ImmutableList.Create("asdasdasd")));
+            Assert.AreEqual("did not recognize '2020-03-22T01:59:20+02' as a UTC-datetime", ex1.Message);
+            Assert.AreEqual("did not recognize 'asdasdasd' as a UTC-datetime", ex2.Message);
+        }
+
+        [Test]
+        public async Task TestOneOfParser()
```

Comment:
might be a good idea to add a case that tests if you supply two items to OneOf

Reply:
how do you supply two items to OneOf? Do you mean using the constructor directly and checking for an exception?

Reply:
supply two items in the immutable list and check for an exception, it should only parse one

Reply:
ideally it should detect if two items were passed and throw because there were on other types given, except a oneof with two generic types

Reply:
I see. I'll add a test for "too many arguments"